src/vcBPA.cpp

```diff
@@ -792,97 +732,126 @@ size_t vcBPA_FindClosestTriangle(vcBPAGrid *pOldGrid, udDouble3 position)
-  udDouble3 aabbExtents = udDouble3::create(itemData.pManifold->gridSize / 2.0);
-  aabbExtents += udDouble3::create(2 * itemData.pManifold->ballRadius); // Add overlap
+      for (int i = 0; i < 3; ++i)
+      {
+        minPos[i] = minPos[i] - udMod(minPos[i], pData->pData->gridSize);
+        maxPos[i] = maxPos[i] + pData->pData->gridSize - udMod(maxPos[i], pData->pData->gridSize);
+      }
 
-  bool hasPoints = false;
-  bool hasNeighbours = false;
-  vcBPA_PopulateGrid(itemData.pManifold->pContext, itemData.pOldModel, nullptr, aabbCenter, aabbExtents, &itemData.oldGrid, &hasPoints, &hasNeighbours);
-  vcBPA_DoGrid(&itemData.oldGrid, itemData.pManifold->ballRadius);
+      udDouble3 currPos;
+      for (currPos.x = minPos.x + pData->pData->gridSize * .5; currPos.x < maxPos.x; currPos.x += pData->pData->gridSize)
+      {
+        for (currPos.y = minPos.x + pData->pData->gridSize * .5; currPos.y < maxPos.y; currPos.y += pData->pData->gridSize)
+        {
+          for (currPos.z = minPos.x + pData->pData->gridSize * .5; currPos.z < maxPos.z; currPos.z += pData->pData->gridSize)
+            vcBPA_AddGrid(pData->pData->pManifold, pData->pData->pOldModel, currPos);
+        }
+      }
+    }
+  }
 
-  udSafeDeque_PushBack(pData->pConvertItemData, itemData);
+  vdkPointBufferF64_Destroy(&pBuffer);
 }
 
 uint32_t vcBPA_GridGeneratorThread(void *pDataPtr)
 {
   vcBPAConvertItem *pData = (vcBPAConvertItem *)pDataPtr;
+  vcBPAOctantData *pOctData;
+  vcBPAOctNode node;
 
-  vcBPAGrid *pGrid = nullptr;
-  int i = 0;
-
-  vdkPointCloudHeader header;
-  vdkPointCloud_GetHeader(pData->pNewModel, &header);
-
-  while (vcBPA_GetGrid(pData->pManifold, pData->pNewModel, &header.attributes, &pGrid, false) && pData->running)
+  while (pData->running)
   {
-    vcBPAConvertItemData data = {};
-    data.pManifold = pData->pManifold;
-    data.pGrid = pGrid;
-    data.pointIndex = 0;
-    data.pOldModel = pData->pOldModel;
-
-    udSafeDeque_PushBack(pData->pQueueItems, data);
-    udWorkerPool_AddTask(pData->pManifold->pPool, vcBPA_GridPopulationThread, pData, false);
-    ++i;
-
-    // Catch up to avoid running out of memory
-    if (i >= 100)
+    if (udSafeDeque_PopBack(pData->pLoadList, &node) != udR_Success)
     {
-      do
-      {
-        udLockMutex(pData->pQueueItems->pMutex);
-        i = (int)pData->pQueueItems->chunkedArray.length;
-        udReleaseMutex(pData->pQueueItems->pMutex);
+      udSleep(200);
+      continue;
+    }
 
-        if (i < 50)
-          break;
+    for (int i = 0; i < 8; ++i)
+    {
+      pOctData = udAllocType(vcBPAOctantData, 1, udAF_Zero);
+      pOctData->pData = pData;
+      pOctData->ppNode = &node.pChildren + i;
+      pOctData->extents = node.extents * .5;
+      pOctData->center = node.center + (octantPositions[i] * pOctData->extents);
 
-        udSleep(500);
-      } while (pData->running);
+      udWorkerPool_AddTask(pData->pManifold->pPool, vcBPA_ProcessOctant, pOctData, true);
```

Comment:
As I was suggesting above with the capturing lambda, you could change this block to be:
```c++
vcBPAOctNode **ppNode = &node.pChildren + i;
udDouble3 extents = node.extents * 0.5;
udDouble3 center = node.center + (octantPositions[i] * pOctData->extents);

udWorkerPool_AddTask(pData->pManifold->pPool, [pData, ppNode, extents, center](void*) { vcBPA_ProcessOctant(pData, ppNode, extents, center); }, nullptr, true);
```
You can probably change that `true` to a `false` - I can't remember what it's for - and then omit it entirely I think?

Also, you might be able to pass `node` by value, and then use the lambda to clean it up. If you do this, it's probably best to put the lambda on its own line since it's going to be multiple lines long (I don't recall how we do this in other places, it might be done within the function call itself rather than as another variable):
```c++
auto callback = […](…) { // or udCallback<void(void*)>
  ...
};
udWorkerPool_AddTask(…, callback, …);
```